app/javascript/components/Admin/Products/Purchases/index.tsx

```diff
@@ -35,6 +45,79 @@ const AdminProductPurchases = ({ productId, isAffiliateUser = false, userId }: A
     mode: "append",
   });
 
+  React.useEffect(() => {
+    if (!open) {
+      setSelectedPurchaseIds(new Set());
+    }
+  }, [open]);
+
+  React.useEffect(() => {
+    setSelectedPurchaseIds((prev) => {
+      const allowedIds = new Set(purchases.map((purchase) => purchase.id));
+      const next = new Set<number>();
+      prev.forEach((id) => {
+        if (allowedIds.has(id)) next.add(id);
+      });
+      return next.size === prev.size ? prev : next;
+    });
```

Comment:
Maybe something like

```suggestion
    setSelectedPurchaseIds((prev) => prev.filter(
      (id) => purchases.some((purchase) => purchase.id === id)
    ))
```

Note: This is an admin page and we don't expect a vast number of purchases so we can prioritize simplicity over performance